<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

## Introduction

In this notebook, we will use the Foursquare API to explore neighborhoods in Toronto. We will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters using the *k*-means clustering algorithm. We will also use the Folium library to visualize the neighborhoods in Toronto and their emerging clusters.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a><br/>

2. <a href="#item2">Explore Neighborhoods in North York, Toronto</a><br/>

3. <a href="#item3">Analyze Each Neighborhood</a><br/>

4. <a href="#item4">Cluster Neighborhoods</a><br/>

5. <a href="#item5">Examine Clusters</a><br/>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [6]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

## 1. Download and Explore Dataset

The first thing we require is a list of neighbourhoods in Toronto. One such list can be obtained from the Wikipedia page: [List of Postal Codes of Canada: M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M). We can copy this list from the wikipedia page and create a csv file. An alternate would be to use web scraping libraries like Beautiful-Soap (however it seems like an overkill for the task)

In [7]:
toronto_codes = pd.read_csv("toronto_codes.csv")
toronto_codes

Postcode           Borough  \
0        M1A      Not assigned   
1        M1B       Scarborough   
2        M1B       Scarborough   
3        M1C       Scarborough   
4        M1C       Scarborough   
5        M1C       Scarborough   
6        M1E       Scarborough   
7        M1E       Scarborough   
8        M1E       Scarborough   
9        M1G       Scarborough   
10       M1H       Scarborough   
11       M1J       Scarborough   
12       M1K       Scarborough   
13       M1K       Scarborough   
14       M1K       Scarborough   
15       M1L       Scarborough   
16       M1L       Scarborough   
17       M1L       Scarborough   
18       M1M       Scarborough   
19       M1M       Scarborough   
20       M1M       Scarborough   
21       M1N       Scarborough   
22       M1N       Scarborough   
23       M1P       Scarborough   
24       M1P       Scarborough   
25       M1P       Scarborough   
26       M1R       Scarborough   
27       M1R       Scarborough   
28       M1S       Scarborough   
29       M1T       Scarborough   
30       M1T       Scarborough   
31       M1T       Scarborough   
32       M1V       Scarborough   
33       M1V       Scarborough   
34       M1V       Scarborough   
35       M1V       Scarborough   
36       M1W       Scarborough   
37       M1W       Scarborough   
38       M1X       Scarborough   
39       M1Y      Not assigned   
40       M1Z      Not assigned   
41       M2A      Not assigned   
42       M2B      Not assigned   
43       M2C      Not assigned   
44       M2E      Not assigned   
45       M2G      Not assigned   
46       M2H        North York   
47       M2J        North York   
48       M2J        North York   
49       M2J        North York   
50       M2K        North York   
51       M2L        North York   
52       M2L        North York   
53       M2M        North York   
54       M2M        North York   
55       M2N        North York   
56       M2P        North York   
57       M2R        North York   
58       M2S      Not assigned   
59       M2T      Not assigned   
60       M2V      Not assigned   
61       M2W      Not assigned   
62       M2X      Not assigned   
63       M2Y      Not assigned   
64       M2Z      Not assigned   
65       M3A        North York   
66       M3B        North York   
67       M3C        North York   
68       M3C        North York   
69       M3E      Not assigned   
70       M3G      Not assigned   
71       M3H        North York   
72       M3H        North York   
73       M3H        North York   
74       M3J        North York   
75       M3J        North York   
76       M3K        North York   
77       M3K        North York   
78       M3L        North York   
79       M3M        North York   
80       M3N        North York   
81       M3P      Not assigned   
82       M3R      Not assigned   
83       M3S      Not assigned   
84       M3T      Not assigned   
85       M3V      Not assigned   
86       M3W      Not assigned   
87       M3X      Not assigned   
88       M3Y      Not assigned   
89       M3Z      Not assigned   
90       M4A        North York   
91       M4B         East York   
92       M4B         East York   
93       M4C         East York   
94       M4E      East Toronto   
95       M4G         East York   
96       M4H         East York   
97       M4J         East York   
98       M4K      East Toronto   
99       M4K      East Toronto   
100      M4L      East Toronto   
101      M4L      East Toronto   
102      M4M      East Toronto   
103      M4N   Central Toronto   
104      M4P   Central Toronto   
105      M4R   Central Toronto   
106      M4S   Central Toronto   
107      M4T   Central Toronto   
108      M4T   Central Toronto   
109      M4V   Central Toronto   
110      M4V   Central Toronto   
111      M4V   Central Toronto   
112      M4V   Central Toronto   
113      M4V   Central Toronto   
114      M4W  Downtown Toronto   
115      M4X  Downtown Toronto   
116      M4X  Downtown Tor

**Preprocessing the codes**

We notice in the above table, that there are several boroughs that are not assigned. We drop those rows.

In [16]:
toronto_codes.drop(toronto_codes[toronto_codes['Borough'] == 'Not assigned'].index, axis=0, inplace=True)

In [17]:
toronto_codes

Postcode           Borough  \
2        M1B       Scarborough   
3        M1C       Scarborough   
4        M1C       Scarborough   
5        M1C       Scarborough   
6        M1E       Scarborough   
7        M1E       Scarborough   
8        M1E       Scarborough   
9        M1G       Scarborough   
10       M1H       Scarborough   
11       M1J       Scarborough   
12       M1K       Scarborough   
13       M1K       Scarborough   
14       M1K       Scarborough   
15       M1L       Scarborough   
16       M1L       Scarborough   
17       M1L       Scarborough   
18       M1M       Scarborough   
19       M1M       Scarborough   
20       M1M       Scarborough   
21       M1N       Scarborough   
22       M1N       Scarborough   
23       M1P       Scarborough   
24       M1P       Scarborough   
25       M1P       Scarborough   
26       M1R       Scarborough   
27       M1R       Scarborough   
28       M1S       Scarborough   
29       M1T       Scarborough   
30       M1T       Scarborough   
31       M1T       Scarborough   
32       M1V       Scarborough   
33       M1V       Scarborough   
34       M1V       Scarborough   
35       M1V       Scarborough   
36       M1W       Scarborough   
37       M1W       Scarborough   
38       M1X       Scarborough   
46       M2H        North York   
47       M2J        North York   
48       M2J        North York   
49       M2J        North York   
50       M2K        North York   
51       M2L        North York   
52       M2L        North York   
53       M2M        North York   
54       M2M        North York   
55       M2N        North York   
56       M2P        North York   
57       M2R        North York   
65       M3A        North York   
66       M3B        North York   
67       M3C        North York   
68       M3C        North York   
71       M3H        North York   
72       M3H        North York   
73       M3H        North York   
74       M3J        North York   
75       M3J        North York   
76       M3K        North York   
77       M3K        North York   
78       M3L        North York   
79       M3M        North York   
80       M3N        North York   
90       M4A        North York   
91       M4B         East York   
92       M4B         East York   
93       M4C         East York   
94       M4E      East Toronto   
95       M4G         East York   
96       M4H         East York   
97       M4J         East York   
98       M4K      East Toronto   
99       M4K      East Toronto   
100      M4L      East Toronto   
101      M4L      East Toronto   
102      M4M      East Toronto   
103      M4N   Central Toronto   
104      M4P   Central Toronto   
105      M4R   Central Toronto   
106      M4S   Central Toronto   
107      M4T   Central Toronto   
108      M4T   Central Toronto   
109      M4V   Central Toronto   
110      M4V   Central Toronto   
111      M4V   Central Toronto   
112      M4V   Central Toronto   
113      M4V   Central Toronto   
114      M4W  Downtown Toronto   
115      M4X  Downtown Toronto   
116      M4X  Downtown Toronto   
117      M4Y  Downtown Toronto   
119      M5A  Downtown Toronto   
120      M5A  Downtown Toronto   
121      M5B  Downtown Toronto   
122      M5B  Downtown Toronto   
123      M5C  Downtown Toronto   
124      M5E  Downtown Toronto   
125      M5G  Downtown Toronto   
126      M5H  Downtown Toronto   
127      M5H  Downtown Toronto   
128      M5H  Downtown Toronto   
129      M5J  Downtown Toronto   
130      M5J  Downtown Toronto   
131      M5J  Downtown Toronto   
132      M5K  Downtown Toronto   
133      M5K  Downtown Toronto   
134      M5L  Downtown Toronto   
135      M5L  Downtown Toronto   
136      M5M        North York   
137      M5M        North York   
138      M5N   Central Toronto   
139      M5P   Central Toronto   
140      M5P   Central Toronto   
141      M5R   Central Toronto   
142      M5R   Central Toronto   
143      M5R   Central Toronto   
144      M5S  Downtown Tor

In [19]:
toronto_codes[toronto_codes['Neighbourhood'] == 'Not assigned']

Postcode       Borough Neighbourhood
195      M7A  Queen's Park  Not assigned

We also see that there still some neighbourhoods that have not been assigned. For these, we assign the neighbourbood the same value of the borough

In [27]:
not_assigned_neighbourhoods = toronto_codes[toronto_codes['Neighbourhood'] == 'Not assigned'].index
toronto_codes.loc[not_assigned_neighbourhoods, 'Neighbourhood'] = toronto_codes.loc[not_assigned_neighbourhoods, 'Borough']

In [30]:
toronto_codes.loc[not_assigned_neighbourhoods]

Postcode       Borough Neighbourhood
195      M7A  Queen's Park  Queen's Park

In [49]:
toronto_codes.shape

(211, 3)

In [48]:
toronto_codes['Postcode'].unique().shape

(103,)

We can see that Toronto has 103 boroughs comprising of 211 neighbourhoods. We can now find the neighbourhoods for each borough

In [67]:
neighbourhoods = pd.DataFrame(toronto_codes.groupby(['Postcode', 'Borough'], axis=0).apply(lambda x : ','.join(x['Neighbourhood'])))
neighbourhoods.reset_index(inplace=True)
neighbourhoods.rename(columns = {0: "Neighbourhood"}, inplace=True)
neighbourhoods

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood  
0                                              Malvern  
1                 Highland Creek,Rouge Hill,Port Union  
2                      Guildwood,Morningside,West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6            East Birchmount Park,Ionview,Kennedy Park  
7          

**Finding co-ordinates of Boroughs in Toronto**

To find the latitude and longitude of the boroughs we can use the file from [http://cocl.us/Geospatial_data](http://cocl.us/Geospatial_data)

In [86]:
!wget -q -O 'toronto_geospatial_data.csv' http://cocl.us/Geospatial_data
print('Data downloaded!')

Data downloaded!


#### Load and explore the geospatial data

Next, let's load the data.

In [87]:
geospatial_data = pd.read_csv('toronto_geospatial_data.csv')
geospatial_data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
30          M3K  43.737473 -79.464763
31          M3L  43.739015 -79.506944
32          M3M  43.728496 -79.495697
33          M3N  43.761631 -79.520999
34          M4A  43.725882 -79.315572
35          M4B  43.706397 -79.309937
36          M4C  43.695344 -79.318389
37          M4E  43.676357 -79.293031
38          M4G  43.709060 -79.363452
39          M4H  43.705369 -79.349372
40          M4J  43.685347 -79.338106
41          M4K  43.679557 -79.352188
42          M4L  43.668999 -79.315572
43          M4M  43.659526 -79.340923
44          M4N  43.728020 -79.388790
45          M4P  43.712751 -79.390197
46          M4R  43.715383 -79.405678
47          M4S  43.704324 -79.388790
48          M4T  43.689574 -79.383160
49          M4V  43.686412 -79.400049
50          M4W  43.679563 -79.377529
51          M4X  43.667967 -79.367675
52          M4Y  43.665860 -79.383160
53          M5A  43.654260 -79.360636
54          M5B  43.657162 -79.378937
55          M5C  43.651494 -79.375418
56          M5E  43.644771 -79.373306
57          M5G  43.657952 -79.387383
58          M5H  43.650571 -79.384568
59          M5J  43.640816 -79.381752
60          M5K  43.647177 -79.381576
61          M5L  43.648198 -79.379817
62          M5M  43.733283 -79.419750
63          M5N  43.711695 -79.416936
64          M5P  43.696948 -79.411307
65          M5R  43.672710 -79.405678
66          M5S  43.662696 -79.400049
67          M5T  43.653206 -79.400049
68          M5V  43.628947 -79.394420
69          M5W  43.646435 -79.374846
70          M5X  43.648429 -79.382280
71          M6A  43.718518 -79.464763
72          M6B  43.709577 -79.445073
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

We can create a single dataframe by joining the <i>neighbourhoods</i> and <i>geospatial_data</i> dataframes on the postal code

In [91]:
neighbourhoods.set_index('Postcode', inplace=True)

In [92]:
geospatial_data.set_index('Postal Code', inplace=True)

In [106]:
neighborhoods = neighbourhoods.join(geospatial_data, how="inner")
neighborhoods.reset_index(inplace=True)
neighborhoods

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                              Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.23

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [107]:
print('The dataframe has {} postal codes.'.format(
        len(neighborhoods['Postcode'].unique())
    )
)

The dataframe has 103 postal codes.


#### Use geopy library to get the latitude and longitude values of Toronto.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>toronto_explorer</em>, as shown below.

In [108]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [117]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_newyork

We will further investigate the North York borough

In [114]:
north_york_data = neighborhoods[neighborhoods['Borough'] == 'North York'].reset_index(drop=True)
north_york_data

Postcode     Borough                                  Neighbourhood  \
0       M2H  North York                              Hillcrest Village   
1       M2J  North York                     Fairview,Henry Farm,Oriole   
2       M2K  North York                                Bayview Village   
3       M2L  North York                        Silver Hills,York Mills   
4       M2M  North York                         Newtonbrook,Willowdale   
5       M2N  North York                               Willowdale South   
6       M2P  North York                                York Mills West   
7       M2R  North York                                Willowdale West   
8       M3A  North York                                      Parkwoods   
9       M3B  North York                                Don Mills North   
10      M3C  North York                Flemingdon Park,Don Mills South   
11      M3H  North York  Bathurst Manor,Downsview North,Wilson Heights   
12      M3J  North York                 Northwood Park,York University   
13      M3K  North York                     CFB Toronto,Downsview East   
14      M3L  North York                                 Downsview West   
15      M3M  North York                              Downsview Central   
16      M3N  North York                            Downsview Northwest   
17      M4A  North York                               Victoria Village   
18      M5M  North York               Bedford Park,Lawrence Manor East   
19      M6A  North York                Lawrence Heights,Lawrence Manor   
20      M6B  North York                                      Glencairn   
21      M6L  North York         Maple Leaf Park,North Park,Upwood Park   
22      M9L  North York                                  Humber Summit   
23      M9M  North York                                Emery,Humberlea   

     Latitude  Longitude  
0   43.803762 -79.363452  
1   43.778517 -79.346556  
2   43.786947 -79.385975  
3   43.757490 -79.374714  
4   43.789053 -79.408493  
5   43.770120 -79.408493  
6   43.752758 -79.400049  
7   43.782736 -79.442259  
8   43.753259 -79.329656  
9   43.745906 -79.352188  
10  43.725900 -79.340923  
11  43.754328 -79.442259  
12  43.767980 -79.487262  
13  43.737473 -79.464763  
14  43.739015 -79.506944  
15  43.728496 -79.495697  
16  43.761631 -79.520999  
17  43.725882 -79.315572  
18  43.733283 -79.419750  
19  43.718518 -79.464763  
20  43.709577 -79.445073  
21  43.713756 -79.490074  
22  43.756303 -79.565963  
23  43.724766 -79.532242

Let's get the geographical coordinates of North York, Toronto, Ontario.

In [115]:
address = 'North York, Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7708175, -79.4132998.


As we did with all of Toronto, let's visualizat North York the neighborhoods in it.

In [120]:
# create map of North York using latitude and longitude values
map_north_york = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(north_york_data['Latitude'], north_york_data['Longitude'], north_york_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_north_york)  
    
map_north_york

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [1]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20190301' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 
CLIENT_SECRET:


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [123]:
north_york_data.loc[0, 'Neighbourhood']

'Hillcrest Village'

Get the neighborhood's latitude and longitude values.

In [124]:
neighborhood_latitude = north_york_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = north_york_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = north_york_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Hillcrest Village are 43.8037622, -79.3634517.


#### Now, let's get the top 100 venues that are within a radius of 1km from Hillcrest Village

First, let's create the GET request URL. Name your URL **url**.

In [125]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url


'https://api.foursquare.com/v2/venues/explore?&client_id=3XK1UBWEBDUCTRUVD5G51DEIH2B2KJLULMKAD4XV4UG01RC5&client_secret=0R4324WC0IZQOMWTMDVXP4JQRONDDNEWJIQ0P421R3CIYJDA&v=20190301&ll=43.8037622,-79.3634517&radius=1000&limit=100'

Send the GET request and examine the resutls

In [126]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c7a79e4dd57977bd317c121'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4bd9842be914a593adbd56fa-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16a941735',
         'name': 'Bakery',
         'pluralName': 'Bakeries',
         'primary': True,
         'shortName': 'Bakery'}],
       'id': '4bd9842be914a593adbd56fa',
       'location': {'address': '3913 Don Mills Rd.',
        'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'at Cliffwood Rd.',
        'distance': 692,
        'formattedAddress': ['3913 Don Mills Rd. (at Cliffwood Rd.)',
         'Toronto ON M2H 2S7',
         'Canada'],
        'labeledLatLngs': [

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [127]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [128]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat  \
0                                        Tastee             Bakery  43.807722   
1  고려삼계탕 Korean Ginseng Chicken Soup & Bibimbap  Korean Restaurant  43.798391   
2                                        Galati      Grocery Store  43.797831   
3                                   Cummer Park               Park  43.799564   
4                                   Tim Hortons        Coffee Shop  43.798945   

         lng  
0 -79.356798  
1 -79.369187  
2 -79.369410  
3 -79.371175  
4 -79.369644

And how many venues were returned by Foursquare?

In [129]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

21 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in North York

#### Let's create a function to repeat the same process to all the neighborhoods in North York

In [130]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [132]:
north_york_venues = getNearbyVenues(names=north_york_data['Neighbourhood'],
                                   latitudes=north_york_data['Latitude'],
                                   longitudes=north_york_data['Longitude']
                                  )



Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Bedford Park,Lawrence Manor East
Lawrence Heights,Lawrence Manor
Glencairn
Maple Leaf Park,North Park,Upwood Park
Humber Summit
Emery,Humberlea


#### Let's check the size of the resulting dataframe

In [134]:
print(north_york_venues.shape)
north_york_venues.head()

(240, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0           Hillcrest Village              43.803762              -79.363452   
1           Hillcrest Village              43.803762              -79.363452   
2           Hillcrest Village              43.803762              -79.363452   
3           Hillcrest Village              43.803762              -79.363452   
4  Fairview,Henry Farm,Oriole              43.778517              -79.346556   

                    Venue  Venue Latitude  Venue Longitude  \
0  Eagle's Nest Golf Club       43.805455       -79.364186   
1         AY Jackson Pool       43.804515       -79.366138   
2            Villa Madina       43.801685       -79.363938   
3       Duncan Creek Park       43.805539       -79.360695   
4          The LEGO Store       43.778207       -79.343483   

             Venue Category  
0               Golf Course  
1                      Pool  
2  Mediterranean Restaurant  
3                   Dog Run  
4          Toy / Game Store

Let's check how many venues were returned for each neighborhood

In [139]:
north_york_venues[['Neighborhood', 'Venue']].groupby('Neighborhood').count()

Venue
Neighborhood                                        
Bathurst Manor,Downsview North,Wilson Heights     18
Bayview Village                                    4
Bedford Park,Lawrence Manor East                  24
CFB Toronto,Downsview East                         3
Don Mills North                                    5
Downsview Central                                  3
Downsview Northwest                                4
Downsview West                                     4
Emery,Humberlea                                    1
Fairview,Henry Farm,Oriole                        69
Flemingdon Park,Don Mills South                   21
Glencairn                                          5
Hillcrest Village                                  4
Humber Summit                                      2
Lawrence Heights,Lawrence Manor                   14
Maple Leaf Park,North Park,Upwood Park             4
Northwood Park,York University                     5
Parkwoods                                          3
Victoria Village                                   5
Willowdale South                                  35
Willowdale West                                    4
York Mills West                                    3

#### Let's find out how many unique categories can be curated from all the returned venues

In [141]:
print('There are {} uniques categories.'.format(len(north_york_venues['Venue Category'].unique())))

There are 108 uniques categories.


In [ ]:
# rename Neighbourhood to Neighborhood
north_york_data.rename(columns={"Neighbourhood" : "Neighborhood"}, inplace=True)

Let's see which neighborhoods do not have any venues

In [175]:
north_york_data.loc[~north_york_data['Neighborhood'].isin(north_york_venues['Neighborhood'].unique())]

Postcode     Borough             Neighborhood   Latitude  Longitude
3      M2L  North York  Silver Hills,York Mills  43.757490 -79.374714
4      M2M  North York   Newtonbrook,Willowdale  43.789053 -79.408493

<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [143]:
# one hot encoding
north_york_onehot = pd.get_dummies(north_york_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
north_york_onehot['Neighborhood'] = north_york_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [north_york_onehot.columns[-1]] + list(north_york_onehot.columns[:-1])
north_york_onehot = north_york_onehot[fixed_columns]

north_york_onehot.head()

Neighborhood  Accessories Store  Airport  \
0           Hillcrest Village                  0        0   
1           Hillcrest Village                  0        0   
2           Hillcrest Village                  0        0   
3           Hillcrest Village                  0        0   
4  Fairview,Henry Farm,Oriole                  0        0   

   American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0                    0                    0                 0   
1                    0                    0                 0   
2                    0                    0                 0   
3                    0                    0                 0   
4                    0                    0                 0   

   Athletics & Sports  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0                   0       0     0    0               0                 0   
1                   0       0     0    0               0                 0   
2                   0       0     0    0               0                 0   
3                   0       0     0    0               0                 0   
4                   0       0     0    0               0                 0   

   Beer Store  Bike Shop  Boutique  Bridal Shop  Bubble Tea Shop  \
0           0          0         0            0                0   
1           0          0         0            0                0   
2           0          0         0            0                0   
3           0          0         0            0                0   
4           0          0         0            0                0   

   Burger Joint  Burrito Place  Bus Station  Bus Stop  Butcher  Café  \
0             0              0            0         0        0     0   
1             0              0            0         0        0     0   
2             0              0            0         0        0     0   
3             0              0            0         0        0     0   
4             0              0            0         0        0     0   

   Candy Store  Caribbean Restaurant  Chinese Restaurant  Clothing Store  \
0            0                     0                   0               0   
1            0                     0                   0               0   
2            0                     0                   0               0   
3            0                     0                   0               0   
4            0                     0                   0               0   

   Coffee Shop  Comfort Food Restaurant  Construction & Landscaping  \
0            0                        0                           0   
1            0                        0                           0   
2            0                        0                           0   
3            0                        0                           0   
4            0                        0                           0   

   Convenience Store  Cosmetics Shop  Deli / Bodega  Department Store  \
0                  0               0              0                 0   
1                  0               0              0                 0   
2                  0               0              0                 0   
3                  0               0              0                 0   
4                  0               0              0                 0   

   Dessert Shop  Dim Sum Restaurant  Diner  Discount Store  Dog Run  \
0             0                   0      0               0        0   
1             0                   0      0               0        0   
2             0                   0      0               0        0   
3             0                   0      0               0        1   
4             0                   0      0               0        0   

   Electronics Store  Empanada Restaurant  Event Space  Fast Food Restaurant  \
0                  0                    0            0                     0   
1                  0                  

And let's examine the new dataframe size.

In [145]:
north_york_onehot.shape

(240, 109)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [146]:
north_york_grouped = north_york_onehot.groupby('Neighborhood').mean().reset_index()
north_york_grouped

Neighborhood  Accessories Store  \
0   Bathurst Manor,Downsview North,Wilson Heights           0.000000   
1                                 Bayview Village           0.000000   
2                Bedford Park,Lawrence Manor East           0.000000   
3                      CFB Toronto,Downsview East           0.000000   
4                                 Don Mills North           0.000000   
5                               Downsview Central           0.000000   
6                             Downsview Northwest           0.000000   
7                                  Downsview West           0.000000   
8                                 Emery,Humberlea           0.000000   
9                      Fairview,Henry Farm,Oriole           0.014493   
10                Flemingdon Park,Don Mills South           0.000000   
11                                      Glencairn           0.000000   
12                              Hillcrest Village           0.000000   
13                                  Humber Summit           0.000000   
14                Lawrence Heights,Lawrence Manor           0.071429   
15         Maple Leaf Park,North Park,Upwood Park           0.000000   
16                 Northwood Park,York University           0.000000   
17                                      Parkwoods           0.000000   
18                               Victoria Village           0.000000   
19                               Willowdale South           0.000000   
20                                Willowdale West           0.000000   
21                                York Mills West           0.000000   

     Airport  American Restaurant  Arts & Crafts Store  Asian Restaurant  \
0   0.000000             0.000000             0.000000          0.000000   
1   0.000000             0.000000             0.000000          0.000000   
2   0.000000             0.041667             0.000000          0.000000   
3   0.333333             0.000000             0.000000          0.000000   
4   0.000000             0.000000             0.000000          0.000000   
5   0.000000             0.000000             0.000000          0.000000   
6   0.000000             0.000000             0.000000          0.000000   
7   0.000000             0.000000             0.000000          0.000000   
8   0.000000             0.000000             0.000000          0.000000   
9   0.000000             0.014493             0.000000          0.028986   
10  0.000000             0.000000             0.000000          0.095238   
11  0.000000             0.000000             0.000000          0.000000   
12  0.000000             0.000000             0.000000          0.000000   
13  0.000000             0.000000             0.000000          0.000000   
14  0.000000             0.000000             0.000000          0.000000   
15  0.000000             0.000000             0.000000          0.000000   
16  0.000000             0.000000             0.000000          0.000000   
17  0.000000             0.000000             0.000000          0.000000   
18  0.000000             0.000000             0.000000          0.000000   
19  0.000000             0.000000             0.028571          0.000000   
20  0.000000             0.000000             0.000000          0.000000   
21  0.000000             0.000000             0.000000          0.000000   

    Athletics & Sports    Bakery      Bank  Bar  Baseball Field  \
0                 0.00  0.000000  0.055556  0.0        0.000000   
1                 0.00  0.000000  0.250000  0.0        0.000000   
2                 0.00  0.000000  0.000000  0.0        0.000000   
3                 0.00  0.000000  0.000000  0.0        0.000000   
4                 0.00  0.000000  0.000000  0.0        0.000000   
5                 0.00  0.000000  0.000000  0.0        0.333333   
6                 0.25  0.000000  0.000000  0.0        0.000000   
7                 0.00  0.000000  0.250000  0.0        0.000000   
8                 0.00  0.00

#### Let's confirm the new size

In [148]:
north_york_grouped.shape

(22, 109)

#### Let's print each neighborhood along with the top 5 most common venues

In [150]:
num_top_venues = 5

for hood in north_york_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = north_york_grouped[north_york_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bathurst Manor,Downsview North,Wilson Heights----
           venue  freq
0    Coffee Shop  0.11
1    Pizza Place  0.06
2      Gift Shop  0.06
3  Deli / Bodega  0.06
4       Pharmacy  0.06


----Bayview Village----
                 venue  freq
0   Chinese Restaurant  0.25
1                 Bank  0.25
2                 Café  0.25
3  Japanese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park,Lawrence Manor East----
                  venue  freq
0  Fast Food Restaurant  0.08
1           Coffee Shop  0.08
2    Italian Restaurant  0.08
3      Sushi Restaurant  0.08
4                   Pub  0.04


----CFB Toronto,Downsview East----
           venue  freq
0        Airport  0.33
1           Park  0.33
2       Bus Stop  0.33
3       Pharmacy  0.00
4  Movie Theater  0.00


----Don Mills North----
                  venue  freq
0  Gym / Fitness Center   0.2
1  Caribbean Restaurant   0.2
2                  Café   0.2
3   Japanese Restaurant   0.2
4      Basketball Court   0.2


--

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [151]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [152]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = north_york_grouped['Neighborhood']

for ind in np.arange(north_york_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(north_york_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0  Bathurst Manor,Downsview North,Wilson Heights           Coffee Shop   
1                                Bayview Village    Chinese Restaurant   
2               Bedford Park,Lawrence Manor East           Coffee Shop   
3                     CFB Toronto,Downsview East               Airport   
4                                Don Mills North   Japanese Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Sandwich Place                  Bank           Bridal Shop   
1                  Café                  Bank   Japanese Restaurant   
2      Sushi Restaurant  Fast Food Restaurant    Italian Restaurant   
3              Bus Stop                  Park         Women's Store   
4  Gym / Fitness Center  Caribbean Restaurant                  Café   

     5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
0               Restaurant                 Pizza Place         Shopping Mall   
1            Women's Store         Empanada Restaurant        Cosmetics Shop   
2  Comfort Food Restaurant                   Juice Bar                   Pub   
3        Electronics Store  Construction & Landscaping     Convenience Store   
4         Basketball Court         Empanada Restaurant        Cosmetics Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Pharmacy         Deli / Bodega   Fast Food Restaurant  
1         Deli / Bodega      Department Store           Dessert Shop  
2     Indian Restaurant            Restaurant               Pharmacy  
3        Cosmetics Shop         Deli / Bodega       Department Store  
4         Deli / Bodega      Department Store           Dessert Shop

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [153]:
# set number of clusters
kclusters = 5

north_york_grouped_clustering = north_york_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(north_york_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 4, 2, 4, 4, 4, 2, 3, 4], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [176]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

north_york_merged = north_york_data

north_york_merged = north_york_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

north_york_merged.head() # check the last columns!

Postcode     Borough                Neighborhood   Latitude  Longitude  \
0      M2H  North York           Hillcrest Village  43.803762 -79.363452   
1      M2J  North York  Fairview,Henry Farm,Oriole  43.778517 -79.346556   
2      M2K  North York             Bayview Village  43.786947 -79.385975   
3      M2L  North York     Silver Hills,York Mills  43.757490 -79.374714   
4      M2M  North York      Newtonbrook,Willowdale  43.789053 -79.408493   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             4.0               Dog Run           Golf Course   
1             4.0        Clothing Store  Fast Food Restaurant   
2             4.0    Chinese Restaurant                  Café   
3             NaN                   NaN                   NaN   
4             NaN                   NaN                   NaN   

  3rd Most Common Venue     4th Most Common Venue 5th Most Common Venue  \
0                  Pool  Mediterranean Restaurant         Women's Store   
1           Coffee Shop          Asian Restaurant                Bakery   
2                  Bank       Japanese Restaurant         Women's Store   
3                   NaN                       NaN                   NaN   
4                   NaN                       NaN                   NaN   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Construction & Landscaping     Convenience Store        Cosmetics Shop   
1                    Tea Room            Food Court         Women's Store   
2         Empanada Restaurant        Cosmetics Shop         Deli / Bodega   
3                         NaN                   NaN                   NaN   
4                         NaN                   NaN                   NaN   

  9th Most Common Venue 10th Most Common Venue  
0         Deli / Bodega       Department Store  
1      Greek Restaurant      Electronics Store  
2      Department Store           Dessert Shop  
3                   NaN                    NaN  
4                   NaN                    NaN

As noted earlier, "Silver Hills, York Mills" and "Newtonbrook, Willowdale" were neighborhoods with no venues listed. So they have not been assigned to any cluster

Finally, let's visualize the resulting clusters

In [183]:
rainbow[3]

'#ffb360'

In [186]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(north_york_merged['Latitude'], north_york_merged['Longitude'], north_york_merged['Neighborhood'], north_york_merged['Cluster Labels']):
    if(~np.isnan(cluster)):
        label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
        folium.CircleMarker(
            [lat, lon],
            radius=5,
            popup=label,
            color=rainbow[int(cluster-1)],
            fill=True,
            fill_color=rainbow[int(cluster-1)],
            fill_opacity=0.7).add_to(map_clusters)
        
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

#### Cluster 1

In [187]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 0, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
22  North York             0.0   Empanada Restaurant           Pizza Place   

   3rd Most Common Venue       4th Most Common Venue 5th Most Common Venue  \
22     Electronics Store  Construction & Landscaping     Convenience Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22        Cosmetics Shop         Deli / Bodega      Department Store   

   9th Most Common Venue 10th Most Common Venue  
22          Dessert Shop     Dim Sum Restaurant

#### Cluster 2

In [188]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 1, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
19  North York             1.0        Clothing Store  Furniture / Home Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
19              Boutique             Gift Shop           Event Space   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
19    Miscellaneous Shop           Coffee Shop     Accessories Store   

    9th Most Common Venue 10th Most Common Venue  
19  Vietnamese Restaurant                 Bakery

#### Cluster 3

In [189]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 2, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
6   North York             2.0                        Park   
8   North York             2.0                        Park   
13  North York             2.0                     Airport   
14  North York             2.0                        Park   
20  North York             2.0                        Park   
21  North York             2.0  Construction & Landscaping   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Bank         Women's Store     Electronics Store   
8      Food & Drink Shop  Fast Food Restaurant               Dog Run   
13              Bus Stop                  Park         Women's Store   
14         Shopping Mall         Grocery Store                  Bank   
20                Bakery                   Pub   Japanese Restaurant   
21                Bakery      Basketball Court                  Park   

         5th Most Common Venue       6th Most Common Venue  \
6   Construction & Landscaping           Convenience Store   
8      Comfort Food Restaurant  Construction & Landscaping   
13           Electronics Store  Construction & Landscaping   
14  Construction & Landscaping           Convenience Store   
20          Italian Restaurant          Frozen Yogurt Shop   
21               Women's Store           Electronics Store   

   7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
6         Cosmetics Shop               Deli / Bodega      Department Store   
8      Convenience Store              Cosmetics Shop         Deli / Bodega   
13     Convenience Store              Cosmetics Shop         Deli / Bodega   
14        Cosmetics Shop               Deli / Bodega      Department Store   
20   Fried Chicken Joint  Construction & Landscaping     Convenience Store   
21     Convenience Store              Cosmetics Shop         Deli / Bodega   

   10th Most Common Venue  
6            Dessert Shop  
8        Department Store  
13       Department Store  
14           Dessert Shop  
20         Cosmetics Shop  
21       Department Store

#### Cluster 4

In [190]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 3, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
23  North York             3.0        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
23           Event Space     Convenience Store        Cosmetics Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
23         Deli / Bodega      Department Store          Dessert Shop   

   9th Most Common Venue 10th Most Common Venue  
23    Dim Sum Restaurant                  Diner

#### Cluster 5

In [191]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 2, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
6   North York             2.0                        Park   
8   North York             2.0                        Park   
13  North York             2.0                     Airport   
14  North York             2.0                        Park   
20  North York             2.0                        Park   
21  North York             2.0  Construction & Landscaping   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Bank         Women's Store     Electronics Store   
8      Food & Drink Shop  Fast Food Restaurant               Dog Run   
13              Bus Stop                  Park         Women's Store   
14         Shopping Mall         Grocery Store                  Bank   
20                Bakery                   Pub   Japanese Restaurant   
21                Bakery      Basketball Court                  Park   

         5th Most Common Venue       6th Most Common Venue  \
6   Construction & Landscaping           Convenience Store   
8      Comfort Food Restaurant  Construction & Landscaping   
13           Electronics Store  Construction & Landscaping   
14  Construction & Landscaping           Convenience Store   
20          Italian Restaurant          Frozen Yogurt Shop   
21               Women's Store           Electronics Store   

   7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
6         Cosmetics Shop               Deli / Bodega      Department Store   
8      Convenience Store              Cosmetics Shop         Deli / Bodega   
13     Convenience Store              Cosmetics Shop         Deli / Bodega   
14        Cosmetics Shop               Deli / Bodega      Department Store   
20   Fried Chicken Joint  Construction & Landscaping     Convenience Store   
21     Convenience Store              Cosmetics Shop         Deli / Bodega   

   10th Most Common Venue  
6            Dessert Shop  
8        Department Store  
13       Department Store  
14           Dessert Shop  
20         Cosmetics Shop  
21       Department Store